mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp

```diff
@@ -4956,9 +5036,15 @@ LogicalResult PackOp::canonicalize(PackOp packOp, PatternRewriter &rewriter) {
     // Insert a cast if needed
     if (needUpdateDestType) {
       rewriter.setInsertionPointAfter(packOp);
-      auto castOp =
-          rewriter.create<tensor::CastOp>(loc, originalResultType, packOp);
-      rewriter.replaceAllUsesExcept(packOp, castOp, castOp);
+      if (hasTensorSemantics) {
+        auto castOp =
+            rewriter.create<tensor::CastOp>(loc, originalResultType, packOp);
+        rewriter.replaceAllUsesExcept(packOp, castOp, castOp);
+      } else {
+        auto castOp =
+            rewriter.create<memref::CastOp>(loc, originalResultType, packOp);
+        rewriter.replaceAllUsesExcept(packOp, castOp, castOp);
+      }
```

Comment:
nit: we can declare `Operation* castOp` and move the rewrite out from the if-else body.

Reply:
It should be done like
```MLIR
  Operation* castOp;
  if (hasTensorSemantics) {
    castOp = rewriter.create<tensor::CastOp>(loc, originalResultType, packOp);
  } else {
    castOp = rewriter.create<memref::CastOp>(loc, originalResultType, packOp);
  }
  rewriter.replaceAllUsesExcept(packOp, castOp->getResult(0), castOp);
```

is it correct?